pkg/cloudprovider/aws/fleet/vpc.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/patrickmn/go-cache"
+	"go.uber.org/zap"
+)
+
+type VPCProvider struct {
+	launchTemplateProvider *LaunchTemplateProvider
+	subnetProvider         *SubnetProvider
+}
+
+func (p *VPCProvider) GetLaunchTemplate(ctx context.Context, clusterSpec *v1alpha1.ClusterSpec) (*ec2.LaunchTemplate, error) {
+	return p.launchTemplateProvider.Get(ctx, clusterSpec)
+}
+
+func (p *VPCProvider) GetZones(ctx context.Context, clusterName string) ([]string, error) {
+	zonalSubnets, err := p.subnetProvider.Get(ctx, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	zones := []string{}
+	for zone := range zonalSubnets {
+		zones = append(zones, zone)
+	}
+	return zones, nil
+}
+
+func (p *VPCProvider) GetZonalSubnets(ctx context.Context, constraints *cloudprovider.Constraints, clusterName string) (map[string][]*ec2.Subnet, error) {
+	// 1. Get all subnets
+	zonalSubnets, err := p.subnetProvider.Get(ctx, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("getting zonal subnets, %w", err)
+	}
+	// 2. Return specific subnet if specified.
+	if subnetID, ok := constraints.Topology[cloudprovider.TopologyKeySubnet]; ok {
+		for zone, subnets := range zonalSubnets {
+			for _, subnet := range subnets {
+				if subnetID == *subnet.SubnetId {
+					return map[string][]*ec2.Subnet{zone: {subnet}}, nil
+				}
+			}
+		}
+		return nil, fmt.Errorf("no subnet exists named %s", subnetID)
+	}
+	// 3. Constrain by zones
+	constrainedZones, err := p.getConstrainedZones(ctx, constraints, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("getting zones, %w", err)
+	}
+	constrainedZonalSubnets := map[string][]*ec2.Subnet{}
+	for zone, subnets := range zonalSubnets {
+		for _, constrainedZone := range constrainedZones {
+			if zone == constrainedZone {
+				constrainedZonalSubnets[constrainedZone] = subnets
+			}
+		}
+	}
+	if len(constrainedZonalSubnets) == 0 {
+		return nil, fmt.Errorf("failed to find viable zonal subnet pairing")
+	}
+	return constrainedZonalSubnets, nil
+}
+
+func (p *VPCProvider) GetSubnetIds(ctx context.Context, clusterName string) ([]string, error) {
+	zonalSubnets, err := p.subnetProvider.Get(ctx, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	subnetIds := []string{}
+	for _, subnets := range zonalSubnets {
+		for _, subnet := range subnets {
+			subnetIds = append(subnetIds, *subnet.SubnetId)
+		}
+	}
+	return subnetIds, nil
+}
+
+func (p *VPCProvider) getConstrainedZones(ctx context.Context, constraints *cloudprovider.Constraints, clusterName string) ([]string, error) {
+	// 1. Return zone if specified.
+	if zone, ok := constraints.Topology[cloudprovider.TopologyKeyZone]; ok {
+		return []string{zone}, nil
+	}
+	// 2. Return all zone options
+	zones, err := p.GetZones(ctx, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	return zones, nil
+}
+
+type ZonalSubnets map[string][]*ec2.Subnet
+
+type SubnetProvider struct {
```

Comment:
I'm a little wary of implementing multiple public structs per file if they have behaviors (data only structs are fine, e.g. Packer computes Packings). Otherwise, this pattern tends to muddy the encapsulation of the file and encourage more structs to be added to the file over time. 

I think it's fairly reasonable to simply collapse all of these functions into the VPCProvider.

Reply:
Yeah agree, shall I tackle this as part of a follow up PR?